File: common/src/main/java/net/impleri/mobskills/restrictions/Restriction.java
package net.impleri.mobskills.restrictions;

import net.impleri.playerskills.restrictions.AbstractRestriction;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.EntityType;
import net.minecraft.world.entity.MobSpawnType;
import net.minecraft.world.entity.player.Player;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

public class Restriction extends AbstractRestriction<EntityType<?>> {
    public final EntitySpawnMode spawnMode;

    public final boolean usable;

    public final List<MobSpawnType> includeSpawners;

    public final List<MobSpawnType> excludeSpawners;

    public Restriction(
            EntityType entity,
            Predicate<Player> condition,
            @Nullable List<ResourceLocation> includeDimensions,
            @Nullable List<ResourceLocation> excludeDimensions,
            @Nullable List<ResourceLocation> includeBiomes,
            @Nullable List<ResourceLocation> excludeBiomes,
            @Nullable EntitySpawnMode spawnMode,
            @Nullable Boolean usable,
            @Nullable List<MobSpawnType> includeSpawners,
            @Nullable List<MobSpawnType> excludeSpawners,
            @Nullable EntityType replacement
    ) {
        super(entity, condition, includeDimensions, excludeDimensions, includeBiomes, excludeBiomes, replacement);
        this.spawnMode = (spawnMode != null) ? spawnMode : EntitySpawnMode.ALLOW_ALWAYS;
        this.usable = Boolean.TRUE.equals(usable);
        this.includeSpawners = (includeSpawners == null) ? new ArrayList<>() : includeSpawners;
        this.excludeSpawners = (excludeSpawners == null) ? new ArrayList<>() : excludeSpawners;
    }
}

File: common/src/main/kotlin/net/impleri/mobskills/restrictions/Restriction.kt
package net.impleri.mobskills.restrictions

import net.impleri.mobskills.api.EntitySpawnMode
import net.impleri.playerskills.restrictions.AbstractRestriction
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.EntityType
import net.minecraft.world.entity.MobSpawnType
import net.minecraft.world.entity.player.Player

open class Restriction(
  entity: EntityType<*>,
  condition: (Player) -> Boolean,
  includeDimensions: List<ResourceLocation>?,
  excludeDimensions: List<ResourceLocation>?,
  includeBiomes: List<ResourceLocation>?,
  excludeBiomes: List<ResourceLocation>?,
  spawnMode: EntitySpawnMode?,
  usable: Boolean?,
  includeSpawners: List<MobSpawnType>?,
  excludeSpawners: List<MobSpawnType>?,
  replacement: EntityType<*>?,
) : AbstractRestriction<EntityType<*>>(
  entity,
  condition,
  includeDimensions ?: ArrayList(),
  excludeDimensions ?: ArrayList(),
  includeBiomes ?: ArrayList(),
  excludeBiomes ?: ArrayList(),
  replacement,
) {
  val spawnMode = spawnMode ?: EntitySpawnMode.ALLOW_ALWAYS
  val usable = usable ?: false
  val includeSpawners = includeSpawners ?: ArrayList()
  val excludeSpawners = excludeSpawners ?: ArrayList()
}
